src/Analysis/Ast/Impl/Indexing/SymbolIndex.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.Python.Core;
+using Microsoft.Python.Parsing.Ast;
+
+namespace Microsoft.Python.Analysis.Indexing {
+    internal sealed class SymbolIndex : ISymbolIndex {
+        private readonly ConcurrentDictionary<Uri, IReadOnlyList<HierarchicalSymbol>> _index = new ConcurrentDictionary<Uri, IReadOnlyList<HierarchicalSymbol>>();
+        private bool _empty;
+        
+        public SymbolIndex() {
+            _empty = false;
+        }
+        
+        public bool isNotEmpty() {
+            return _empty;
+        }
+
+
+        public IEnumerable<HierarchicalSymbol> HierarchicalDocumentSymbols(Uri uri)
+            => _index.TryGetValue(uri, out var list) ? list : Enumerable.Empty<HierarchicalSymbol>();
+
+        public IEnumerable<FlatSymbol> WorkspaceSymbols(string query) {
+            foreach (var kvp in _index) {
+                foreach (var found in WorkspaceSymbolsQuery(query, kvp.Key, kvp.Value)) {
+                    yield return found;
+                }
+            }
+        }
+
+        private IEnumerable<FlatSymbol> WorkspaceSymbolsQuery(string query, Uri uri, IEnumerable<HierarchicalSymbol> symbols) {
+            // Some semblance of a BFS.
+            var queue = new Queue<(HierarchicalSymbol, string)>(symbols.Select(s => (s, (string)null)));
```

Comment:
We have `TraverseBreadthFirst` Linq extension.

Reply:
We discussed this yesterday and on the other PR; that extension cannot be used because it assumes a single root.

Reply:
It'd probably be smart to move this along side that other extension with a similar signature.